File: apps/ios/Sources/CrossyUI/GridModule.swift
// The 36-unit cell module (root DESIGN.md §10, Wave 2.1d): clue number top-left
// (+2,+10), teammate presence anchored bottom-right, direction arrow top-right
// (+27,+3, a 7-unit glyph), avatar circle bottom-right (center +30,+30, radius 5,
// 8 px initial), count badge in the same bottom-right slot (center +29,+29, radius 7,
// 9 px count). Those constants are the cross-client module contract; everything here
// is in module units, scaled to points by the camera (GridCamera). Positions written
// as baselines in the spec are converted for Canvas anchor drawing by capCenterY.

import CoreGraphics

public enum GridModule {
    /// The cell module edge, in module units (root DESIGN.md §10: "a 36-unit cell
    /// module scaled to fit").
    public static let unit: CGFloat = 36

    // MARK: Clue number, top-left (Wave 2.1d: +2,+10 baseline, 10-unit type)

    public static let numberLeading: CGFloat = 2
    public static let numberBaseline: CGFloat = 10
    public static let numberFontSize: CGFloat = 10

    // MARK: Entry glyph, centered (web baseline parity: 24-unit type, baseline +32)

    public static let glyphFontSize: CGFloat = 24
    public static let glyphCenterX: CGFloat = unit / 2
    public static let glyphBaseline: CGFloat = 32
    /// The glyph steps 3 units left of center when a presence mark shares the cell,
    /// clearing the bottom-right stack (the web renderer's letterX rule).
    public static let glyphPresenceShift: CGFloat = -3

    // MARK: Circles (root DESIGN.md §10: inset rings)

    public static let circleRadius: CGFloat = unit / 2.1
    public static let circleStroke: CGFloat = 0.8
    /// Shaded circles are a render variant of circles (CrossyProtocol ClientPuzzle):
    /// a soft achromatic wash instead of a ring, alpha over the cell fill.
    public static let shadeAlpha: Double = 0.18

    // MARK: Presence, bottom-right stack (Wave 2.1d)

    public static let arrowOrigin = CGPoint(x: 27, y: 3)
    public static let arrowSize: CGFloat = 7
    public static let avatarCenter = CGPoint(x: 30, y: 30)
    public static let avatarRadius: CGFloat = 5
    public static let avatarInitialFontSize: CGFloat = 8
    public static let badgeCenter = CGPoint(x: 29, y: 29)
    public static let badgeRadius: CGFloat = 7
    public static let badgeCountFontSize: CGFloat = 9

    // MARK: Lines

    /// Cell hairline (web parity: 0.6 stroke) and the closing outer frame (2 units).
    public static let hairline: CGFloat = 0.6
    public static let frameStroke: CGFloat = 2

    // MARK: Cell geometry

    public static func cellOrigin(_ cell: Int, cols: Int) -> CGPoint {
        CGPoint(x: CGFloat(cell % cols) * unit, y: CGFloat(cell / cols) * unit)
    }

    public static func cellRect(_ cell: Int, cols: Int) -> CGRect {
        let origin = cellOrigin(cell, cols: cols)
        return CGRect(x: origin.x, y: origin.y, width: unit, height: unit)
    }

    // MARK: Text placement

    /// The vertical center of a capital glyph drawn on `baseline`. The spec writes
    /// text positions as SVG baselines; Canvas draws anchored text, so the cap-height
    /// midpoint (SF Pro cap height is ~0.72 em) converts one to the other.
    public static func capCenterY(baseline: CGFloat, fontSize: CGFloat) -> CGFloat {
        baseline - fontSize * 0.36
    }

    // MARK: Rebus scaling

    /// The ink width a rebus string may fill, and the SF Pro caps average advance
    /// (~0.62 em) used to fit it. Below `rebusMinimumFontSize` a string stops
    /// shrinking and simply runs tight; a 10-glyph rebus is the charset cap
    /// (PROTOCOL.md §3) and is unreadable at any size in a 36-unit module.
    public static let rebusInkWidth: CGFloat = 32
    public static let rebusCapAdvance: CGFloat = 0.62
    public static let rebusMinimumFontSize: CGFloat = 5

    /// Entry glyph size for a value of `length` characters: 24 units for a single
    /// glyph, longer (rebus) strings scaled to fit the ink width, floored.
    public static func glyphSize(forLength length: Int) -> CGFloat {
        guard length > 1 else { return glyphFontSize }
        let fitted = rebusInkWidth / (rebusCapAdvance * CGFloat(length))
        return min(glyphFontSize, max(rebusMinimumFontSize, fitted))
    }
}

File: apps/android/ui/src/main/kotlin/crossy/ui/GridModule.kt
// The 36-unit cell module (root DESIGN.md §10, Wave 2.1d), the cross-client module contract:
// clue number top-left (+2,+10), teammate presence anchored bottom-right, direction arrow
// top-right (+27,+3), avatar puck bottom-right (center +30,+30, radius 5), count badge in the
// same bottom-right slot (center +29,+29, radius 7). Twin of the iOS GridModule; every value is
// in module units and the grid scales one module to pixels by a single unit factor. These are
// render constants (not wire types), so restating them here is the module contract, not a
// redefinition. Values in Float for Compose draw math.

package crossy.ui

object GridModule {
    /** The cell module edge, in module units (DESIGN.md §10: "a 36-unit cell module scaled to fit"). */
    const val UNIT: Float = 36f

    // Clue number, top-left (Wave 2.1d: +2,+10 baseline, 10-unit type).
    const val NUMBER_LEADING: Float = 2f
    const val NUMBER_BASELINE: Float = 10f
    const val NUMBER_FONT_SIZE: Float = 10f

    // Entry glyph, centered (web baseline parity: 24-unit type).
    const val GLYPH_FONT_SIZE: Float = 24f

    // Circles (DESIGN.md §10: inset rings). Shaded circles are a soft achromatic wash.
    const val CIRCLE_RADIUS: Float = UNIT / 2.1f
    const val CIRCLE_STROKE: Float = 0.8f
    const val SHADE_ALPHA: Float = 0.18f

    // Presence, bottom-right stack (Wave 2.1d).
    const val ARROW_ORIGIN_X: Float = 27f
    const val ARROW_ORIGIN_Y: Float = 3f
    const val ARROW_SIZE: Float = 7f
    const val AVATAR_CENTER_X: Float = 30f
    const val AVATAR_CENTER_Y: Float = 30f
    const val AVATAR_RADIUS: Float = 5f
    const val AVATAR_INITIAL_FONT_SIZE: Float = 8f
    const val BADGE_CENTER_X: Float = 29f
    const val BADGE_CENTER_Y: Float = 29f
    const val BADGE_RADIUS: Float = 7f
    const val BADGE_COUNT_FONT_SIZE: Float = 9f

    // Lines: cell hairline (web parity: 0.6 stroke) and the closing outer frame (2 units).
    const val HAIRLINE: Float = 0.6f
    const val FRAME_STROKE: Float = 2f

    // Background overlay alphas, derived from real tokens (the cursor's roster color), never
    // invented paints: the current cell and its active word are washes of cursorTint so the
    // selection reads at a glance. Tuning is a design-pass concern.
    const val CURRENT_ALPHA: Float = 0.30f
    const val ACTIVE_WORD_ALPHA: Float = 0.13f
    const val TEAMMATE_ALPHA: Float = 0.10f
}
